This notebook implements a chatbot for ICICI Home Finance using Mistral LLM through Ollama.

# 1. Prerequisites


## 1.1 Ollama Installation

In [1]:
# Install ollama
# !curl https://ollama.ai/install.sh | sh
# Start Ollama server
import subprocess
import threading
import time

# Start Ollama in background
def run_ollama():
    subprocess.run(["ollama", "serve"])

ollama_thread = threading.Thread(target=run_ollama)
ollama_thread.start()
time.sleep(5)  # Wait for server initialization


# Pull Mistral
! ollama pull mistral

Error: listen tcp 127.0.0.1:11434: bind: address already in use


pulling manifest ⠋ pulling manifest ⠙ pulling manifest ⠹ pulling manifest ⠸ pulling manifest ⠼ pulling manifest ⠴ pulling manifest ⠦ pulling manifest ⠧ pulling manifest ⠇ pulling manifest ⠏ pulling manifest ⠋ pulling manifest ⠙ pulling manifest ⠹ pulling manifest ⠸ pulling manifest 
pulling ff82381e2bea...   0% ▕                ▏    0 B/4.1 GB                  pulling manifest 
pulling ff82381e2bea...   0% ▕                ▏    0 B/4.1 GB                  pulling manifest 
pulling ff82381e2bea...   0% ▕                ▏    0 B/4.1 GB                  pulling manifest 
pulling ff82381e2bea...   0% ▕                ▏    0 B/4.1 GB                  pulling manifest 
pulling ff82381e2bea...   0% ▕                ▏    0 B/4.1 GB                  pulling manifest 
pulling ff82381e2bea...   0% ▕                ▏    0 B/4.1 GB                  pulling manifest 
pulling ff82381e2bea...   0% ▕                ▏    0 B/4.1 GB                  pulling manifest 
pulling ff82381e2bea...   0% ▕       

pulling ff82381e2bea...   1% ▕                ▏  60 MB/4.1 GB   11 MB/s   5m45spulling manifest 
pulling ff82381e2bea...   1% ▕                ▏  60 MB/4.1 GB   11 MB/s   5m45spulling manifest 
pulling ff82381e2bea...   1% ▕                ▏  60 MB/4.1 GB   10 MB/s   6m38spulling manifest 
pulling ff82381e2bea...   1% ▕                ▏  61 MB/4.1 GB   10 MB/s   6m38spulling manifest 
pulling ff82381e2bea...   1% ▕                ▏  61 MB/4.1 GB   10 MB/s   6m38spulling manifest 
pulling ff82381e2bea...   1% ▕                ▏  61 MB/4.1 GB   10 MB/s   6m38spulling manifest 
pulling ff82381e2bea...   1% ▕                ▏  61 MB/4.1 GB   10 MB/s   6m38spulling manifest 
pulling ff82381e2bea...   2% ▕                ▏  61 MB/4.1 GB   10 MB/s   6m38spulling manifest 
pulling ff82381e2bea...   2% ▕                ▏  62 MB/4.1 GB   10 MB/s   6m38spulling manifest 
pulling ff82381e2bea...   2% ▕                ▏  62 MB/4.1 GB   10 MB/s   6m38spulling manifest 
pulling ff82381e2bea...   2% ▕

pulling ff82381e2bea...   3% ▕                ▏ 129 MB/4.1 GB   10 MB/s   6m18spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 130 MB/4.1 GB   10 MB/s   6m17spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 132 MB/4.1 GB   10 MB/s   6m17spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 134 MB/4.1 GB   10 MB/s   6m17spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 135 MB/4.1 GB   10 MB/s   6m17spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 136 MB/4.1 GB   10 MB/s   6m17spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 138 MB/4.1 GB   10 MB/s   6m17spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 141 MB/4.1 GB   10 MB/s   6m16spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 142 MB/4.1 GB   10 MB/s   6m16spulling manifest 
pulling ff82381e2bea...   3% ▕                ▏ 143 MB/4.1 GB   10 MB/s   6m16spulling manifest 
pulling ff82381e2bea...   4% ▕

pulling ff82381e2bea...   6% ▕                ▏ 227 MB/4.1 GB   15 MB/s    4m7spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 227 MB/4.1 GB   15 MB/s    4m7spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 227 MB/4.1 GB   15 MB/s    4m7spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 229 MB/4.1 GB   15 MB/s    4m7spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 229 MB/4.1 GB   15 MB/s    4m7spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 229 MB/4.1 GB   15 MB/s    4m7spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 229 MB/4.1 GB   15 MB/s    4m7spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 229 MB/4.1 GB   14 MB/s   4m23spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 229 MB/4.1 GB   14 MB/s   4m23spulling manifest 
pulling ff82381e2bea...   6% ▕                ▏ 229 MB/4.1 GB   14 MB/s   4m23spulling manifest 
pulling ff82381e2bea...   6% ▕

pulling ff82381e2bea...   8% ▕█               ▏ 320 MB/4.1 GB   13 MB/s   4m40spulling manifest 
pulling ff82381e2bea...   8% ▕█               ▏ 324 MB/4.1 GB   13 MB/s   4m39spulling manifest 
pulling ff82381e2bea...   8% ▕█               ▏ 328 MB/4.1 GB   13 MB/s   4m39spulling manifest 
pulling ff82381e2bea...   8% ▕█               ▏ 330 MB/4.1 GB   13 MB/s   4m39spulling manifest 
pulling ff82381e2bea...   8% ▕█               ▏ 335 MB/4.1 GB   16 MB/s   3m55spulling manifest 
pulling ff82381e2bea...   8% ▕█               ▏ 339 MB/4.1 GB   16 MB/s   3m55spulling manifest 
pulling ff82381e2bea...   8% ▕█               ▏ 340 MB/4.1 GB   16 MB/s   3m55spulling manifest 
pulling ff82381e2bea...   8% ▕█               ▏ 342 MB/4.1 GB   16 MB/s   3m55spulling manifest 
pulling ff82381e2bea...   8% ▕█               ▏ 348 MB/4.1 GB   16 MB/s   3m54spulling manifest 
pulling ff82381e2bea...   9% ▕█               ▏ 351 MB/4.1 GB   16 MB/s   3m54spulling manifest 
pulling ff82381e2bea...   9% ▕

pulling ff82381e2bea...  14% ▕██              ▏ 571 MB/4.1 GB   33 MB/s   1m45spulling manifest 
pulling ff82381e2bea...  14% ▕██              ▏ 576 MB/4.1 GB   37 MB/s   1m34spulling manifest 
pulling ff82381e2bea...  14% ▕██              ▏ 577 MB/4.1 GB   37 MB/s   1m34spulling manifest 
pulling ff82381e2bea...  14% ▕██              ▏ 580 MB/4.1 GB   37 MB/s   1m34spulling manifest 
pulling ff82381e2bea...  14% ▕██              ▏ 590 MB/4.1 GB   37 MB/s   1m33spulling manifest 
pulling ff82381e2bea...  14% ▕██              ▏ 592 MB/4.1 GB   37 MB/s   1m33spulling manifest 
pulling ff82381e2bea...  14% ▕██              ▏ 594 MB/4.1 GB   37 MB/s   1m33spulling manifest 
pulling ff82381e2bea...  15% ▕██              ▏ 601 MB/4.1 GB   37 MB/s   1m33spulling manifest 
pulling ff82381e2bea...  15% ▕██              ▏ 604 MB/4.1 GB   37 MB/s   1m33spulling manifest 
pulling ff82381e2bea...  15% ▕██              ▏ 609 MB/4.1 GB   37 MB/s   1m33spulling manifest 
pulling ff82381e2bea...  15% ▕

pulling ff82381e2bea...  16% ▕██              ▏ 667 MB/4.1 GB   23 MB/s   2m29spulling manifest 
pulling ff82381e2bea...  16% ▕██              ▏ 669 MB/4.1 GB   23 MB/s   2m28spulling manifest 
pulling ff82381e2bea...  16% ▕██              ▏ 674 MB/4.1 GB   23 MB/s   2m28spulling manifest 
pulling ff82381e2bea...  17% ▕██              ▏ 678 MB/4.1 GB   23 MB/s   2m28spulling manifest 
pulling ff82381e2bea...  17% ▕██              ▏ 681 MB/4.1 GB   23 MB/s   2m28spulling manifest 
pulling ff82381e2bea...  17% ▕██              ▏ 686 MB/4.1 GB   23 MB/s   2m28spulling manifest 
pulling ff82381e2bea...  17% ▕██              ▏ 691 MB/4.1 GB   23 MB/s   2m28spulling manifest 
pulling ff82381e2bea...  17% ▕██              ▏ 694 MB/4.1 GB   23 MB/s   2m27spulling manifest 
pulling ff82381e2bea...  17% ▕██              ▏ 699 MB/4.1 GB   23 MB/s   2m27spulling manifest 
pulling ff82381e2bea...  17% ▕██              ▏ 701 MB/4.1 GB   23 MB/s   2m26spulling manifest 
pulling ff82381e2bea...  17% ▕

pulling ff82381e2bea...  23% ▕███             ▏ 928 MB/4.1 GB   31 MB/s   1m41spulling manifest 
pulling ff82381e2bea...  23% ▕███             ▏ 934 MB/4.1 GB   31 MB/s   1m41spulling manifest 
pulling ff82381e2bea...  23% ▕███             ▏ 940 MB/4.1 GB   31 MB/s   1m41spulling manifest 
pulling ff82381e2bea...  23% ▕███             ▏ 940 MB/4.1 GB   31 MB/s   1m41spulling manifest 
pulling ff82381e2bea...  23% ▕███             ▏ 948 MB/4.1 GB   31 MB/s   1m41spulling manifest 
pulling ff82381e2bea...  23% ▕███             ▏ 950 MB/4.1 GB   31 MB/s   1m41spulling manifest 
pulling ff82381e2bea...  23% ▕███             ▏ 954 MB/4.1 GB   35 MB/s   1m28spulling manifest 
pulling ff82381e2bea...  23% ▕███             ▏ 961 MB/4.1 GB   35 MB/s   1m27spulling manifest 
pulling ff82381e2bea...  23% ▕███             ▏ 962 MB/4.1 GB   35 MB/s   1m27spulling manifest 
pulling ff82381e2bea...  24% ▕███             ▏ 968 MB/4.1 GB   35 MB/s   1m27spulling manifest 
pulling ff82381e2bea...  24% ▕

pulling ff82381e2bea...  25% ▕████            ▏ 1.0 GB/4.1 GB   29 MB/s   1m44spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.0 GB/4.1 GB   29 MB/s   1m44spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.0 GB/4.1 GB   29 MB/s   1m44spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.1 GB/4.1 GB   24 MB/s    2m4spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.1 GB/4.1 GB   24 MB/s    2m4spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.1 GB/4.1 GB   24 MB/s    2m4spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.1 GB/4.1 GB   24 MB/s    2m3spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.1 GB/4.1 GB   24 MB/s    2m3spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.1 GB/4.1 GB   24 MB/s    2m3spulling manifest 
pulling ff82381e2bea...  26% ▕████            ▏ 1.1 GB/4.1 GB   24 MB/s    2m3spulling manifest 
pulling ff82381e2bea...  26% ▕

pulling ff82381e2bea...  31% ▕████            ▏ 1.3 GB/4.1 GB   25 MB/s   1m50spulling manifest 
pulling ff82381e2bea...  31% ▕████            ▏ 1.3 GB/4.1 GB   25 MB/s   1m50spulling manifest 
pulling ff82381e2bea...  31% ▕████            ▏ 1.3 GB/4.1 GB   25 MB/s   1m49spulling manifest 
pulling ff82381e2bea...  31% ▕████            ▏ 1.3 GB/4.1 GB   25 MB/s   1m49spulling manifest 
pulling ff82381e2bea...  31% ▕████            ▏ 1.3 GB/4.1 GB   25 MB/s   1m49spulling manifest 
pulling ff82381e2bea...  31% ▕█████           ▏ 1.3 GB/4.1 GB   25 MB/s   1m49spulling manifest 
pulling ff82381e2bea...  32% ▕█████           ▏ 1.3 GB/4.1 GB   25 MB/s   1m49spulling manifest 
pulling ff82381e2bea...  32% ▕█████           ▏ 1.3 GB/4.1 GB   25 MB/s   1m49spulling manifest 
pulling ff82381e2bea...  32% ▕█████           ▏ 1.3 GB/4.1 GB   25 MB/s   1m49spulling manifest 
pulling ff82381e2bea...  32% ▕█████           ▏ 1.3 GB/4.1 GB   25 MB/s   1m48spulling manifest 
pulling ff82381e2bea...  32% ▕

pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   32 MB/s   1m21spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   32 MB/s   1m21spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   32 MB/s   1m21spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   32 MB/s   1m21spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   32 MB/s   1m21spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   32 MB/s   1m21spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   32 MB/s   1m21spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   32 MB/s   1m21spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   28 MB/s   1m33spulling manifest 
pulling ff82381e2bea...  35% ▕█████           ▏ 1.4 GB/4.1 GB   28 MB/s   1m33spulling manifest 
pulling ff82381e2bea...  35% ▕

pulling ff82381e2bea...  40% ▕██████          ▏ 1.6 GB/4.1 GB   23 MB/s   1m43spulling manifest 
pulling ff82381e2bea...  40% ▕██████          ▏ 1.6 GB/4.1 GB   23 MB/s   1m43spulling manifest 
pulling ff82381e2bea...  40% ▕██████          ▏ 1.6 GB/4.1 GB   23 MB/s   1m43spulling manifest 
pulling ff82381e2bea...  40% ▕██████          ▏ 1.6 GB/4.1 GB   23 MB/s   1m42spulling manifest 
pulling ff82381e2bea...  40% ▕██████          ▏ 1.7 GB/4.1 GB   23 MB/s   1m42spulling manifest 
pulling ff82381e2bea...  40% ▕██████          ▏ 1.7 GB/4.1 GB   24 MB/s   1m40spulling manifest 
pulling ff82381e2bea...  40% ▕██████          ▏ 1.7 GB/4.1 GB   24 MB/s   1m40spulling manifest 
pulling ff82381e2bea...  40% ▕██████          ▏ 1.7 GB/4.1 GB   24 MB/s   1m39spulling manifest 
pulling ff82381e2bea...  41% ▕██████          ▏ 1.7 GB/4.1 GB   24 MB/s   1m39spulling manifest 
pulling ff82381e2bea...  41% ▕██████          ▏ 1.7 GB/4.1 GB   24 MB/s   1m39spulling manifest 
pulling ff82381e2bea...  41% ▕

pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   41 MB/s     54spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   41 MB/s     54spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   37 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   37 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   37 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   37 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   37 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   37 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   37 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  45% ▕███████         ▏ 1.9 GB/4.1 GB   37 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  45% ▕

pulling ff82381e2bea...  48% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m29spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m28spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m28spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m28spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m28spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m28spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m27spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m27spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m27spulling manifest 
pulling ff82381e2bea...  49% ▕███████         ▏ 2.0 GB/4.1 GB   23 MB/s   1m27spulling manifest 
pulling ff82381e2bea...  50% ▕

pulling ff82381e2bea...  55% ▕████████        ▏ 2.3 GB/4.1 GB   41 MB/s     44spulling manifest 
pulling ff82381e2bea...  55% ▕████████        ▏ 2.3 GB/4.1 GB   41 MB/s     44spulling manifest 
pulling ff82381e2bea...  55% ▕████████        ▏ 2.3 GB/4.1 GB   41 MB/s     44spulling manifest 
pulling ff82381e2bea...  55% ▕████████        ▏ 2.3 GB/4.1 GB   41 MB/s     43spulling manifest 
pulling ff82381e2bea...  55% ▕████████        ▏ 2.3 GB/4.1 GB   41 MB/s     43spulling manifest 
pulling ff82381e2bea...  56% ▕████████        ▏ 2.3 GB/4.1 GB   41 MB/s     43spulling manifest 
pulling ff82381e2bea...  56% ▕████████        ▏ 2.3 GB/4.1 GB   41 MB/s     43spulling manifest 
pulling ff82381e2bea...  56% ▕████████        ▏ 2.3 GB/4.1 GB   42 MB/s     43spulling manifest 
pulling ff82381e2bea...  56% ▕████████        ▏ 2.3 GB/4.1 GB   42 MB/s     43spulling manifest 
pulling ff82381e2bea...  56% ▕████████        ▏ 2.3 GB/4.1 GB   42 MB/s     43spulling manifest 
pulling ff82381e2bea...  56% ▕

pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m11spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m11spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m10spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m10spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m10spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m10spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m10spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m10spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m10spulling manifest 
pulling ff82381e2bea...  58% ▕█████████       ▏ 2.4 GB/4.1 GB   24 MB/s   1m10spulling manifest 
pulling ff82381e2bea...  59% ▕

pulling ff82381e2bea...  63% ▕██████████      ▏ 2.6 GB/4.1 GB   30 MB/s     50spulling manifest 
pulling ff82381e2bea...  63% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     44spulling manifest 
pulling ff82381e2bea...  63% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     44spulling manifest 
pulling ff82381e2bea...  63% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     44spulling manifest 
pulling ff82381e2bea...  64% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     44spulling manifest 
pulling ff82381e2bea...  64% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     44spulling manifest 
pulling ff82381e2bea...  64% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     44spulling manifest 
pulling ff82381e2bea...  64% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     44spulling manifest 
pulling ff82381e2bea...  64% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     43spulling manifest 
pulling ff82381e2bea...  64% ▕██████████      ▏ 2.6 GB/4.1 GB   33 MB/s     43spulling manifest 
pulling ff82381e2bea...  64% ▕

pulling ff82381e2bea...  66% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m1spulling manifest 
pulling ff82381e2bea...  66% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m1spulling manifest 
pulling ff82381e2bea...  66% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m1spulling manifest 
pulling ff82381e2bea...  66% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m1spulling manifest 
pulling ff82381e2bea...  66% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m1spulling manifest 
pulling ff82381e2bea...  66% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  66% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  67% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  67% ▕██████████      ▏ 2.7 GB/4.1 GB   22 MB/s    1m0spulling manifest 
pulling ff82381e2bea...  67% ▕██████████      ▏ 2.7 GB/4.1 GB   23 MB/s     57spulling manifest 
pulling ff82381e2bea...  67% ▕

pulling ff82381e2bea...  72% ▕███████████     ▏ 3.0 GB/4.1 GB   31 MB/s     35spulling manifest 
pulling ff82381e2bea...  72% ▕███████████     ▏ 3.0 GB/4.1 GB   31 MB/s     35spulling manifest 
pulling ff82381e2bea...  72% ▕███████████     ▏ 3.0 GB/4.1 GB   31 MB/s     35spulling manifest 
pulling ff82381e2bea...  73% ▕███████████     ▏ 3.0 GB/4.1 GB   31 MB/s     35spulling manifest 
pulling ff82381e2bea...  73% ▕███████████     ▏ 3.0 GB/4.1 GB   31 MB/s     35spulling manifest 
pulling ff82381e2bea...  73% ▕███████████     ▏ 3.0 GB/4.1 GB   31 MB/s     35spulling manifest 
pulling ff82381e2bea...  73% ▕███████████     ▏ 3.0 GB/4.1 GB   36 MB/s     30spulling manifest 
pulling ff82381e2bea...  73% ▕███████████     ▏ 3.0 GB/4.1 GB   36 MB/s     30spulling manifest 
pulling ff82381e2bea...  73% ▕███████████     ▏ 3.0 GB/4.1 GB   36 MB/s     30spulling manifest 
pulling ff82381e2bea...  73% ▕███████████     ▏ 3.0 GB/4.1 GB   36 MB/s     30spulling manifest 
pulling ff82381e2bea...  73% ▕

pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   32 MB/s     30spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   32 MB/s     30spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   32 MB/s     30spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   27 MB/s     35spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   27 MB/s     35spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   27 MB/s     35spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   27 MB/s     35spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   27 MB/s     35spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   27 MB/s     35spulling manifest 
pulling ff82381e2bea...  76% ▕████████████    ▏ 3.1 GB/4.1 GB   27 MB/s     35spulling manifest 
pulling ff82381e2bea...  76% ▕

pulling ff82381e2bea...  81% ▕█████████████   ▏ 3.3 GB/4.1 GB   26 MB/s     29spulling manifest 
pulling ff82381e2bea...  81% ▕█████████████   ▏ 3.3 GB/4.1 GB   26 MB/s     29spulling manifest 
pulling ff82381e2bea...  81% ▕█████████████   ▏ 3.3 GB/4.1 GB   26 MB/s     29spulling manifest 
pulling ff82381e2bea...  82% ▕█████████████   ▏ 3.4 GB/4.1 GB   26 MB/s     29spulling manifest 
pulling ff82381e2bea...  82% ▕█████████████   ▏ 3.4 GB/4.1 GB   26 MB/s     28spulling manifest 
pulling ff82381e2bea...  82% ▕█████████████   ▏ 3.4 GB/4.1 GB   26 MB/s     28spulling manifest 
pulling ff82381e2bea...  82% ▕█████████████   ▏ 3.4 GB/4.1 GB   26 MB/s     28spulling manifest 
pulling ff82381e2bea...  82% ▕█████████████   ▏ 3.4 GB/4.1 GB   26 MB/s     28spulling manifest 
pulling ff82381e2bea...  82% ▕█████████████   ▏ 3.4 GB/4.1 GB   26 MB/s     28spulling manifest 
pulling ff82381e2bea...  82% ▕█████████████   ▏ 3.4 GB/4.1 GB   26 MB/s     28spulling manifest 
pulling ff82381e2bea...  82% ▕

pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   35 MB/s     16spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   35 MB/s     16spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   35 MB/s     16spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   35 MB/s     16spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   35 MB/s     16spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   35 MB/s     16spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   35 MB/s     16spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   35 MB/s     16spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   30 MB/s     19spulling manifest 
pulling ff82381e2bea...  86% ▕█████████████   ▏ 3.5 GB/4.1 GB   30 MB/s     19spulling manifest 
pulling ff82381e2bea...  86% ▕

pulling ff82381e2bea...  90% ▕██████████████  ▏ 3.7 GB/4.1 GB   22 MB/s     19spulling manifest 
pulling ff82381e2bea...  90% ▕██████████████  ▏ 3.7 GB/4.1 GB   22 MB/s     18spulling manifest 
pulling ff82381e2bea...  90% ▕██████████████  ▏ 3.7 GB/4.1 GB   22 MB/s     18spulling manifest 
pulling ff82381e2bea...  90% ▕██████████████  ▏ 3.7 GB/4.1 GB   22 MB/s     18spulling manifest 
pulling ff82381e2bea...  90% ▕██████████████  ▏ 3.7 GB/4.1 GB   22 MB/s     18spulling manifest 
pulling ff82381e2bea...  90% ▕██████████████  ▏ 3.7 GB/4.1 GB   23 MB/s     17spulling manifest 
pulling ff82381e2bea...  90% ▕██████████████  ▏ 3.7 GB/4.1 GB   23 MB/s     17spulling manifest 
pulling ff82381e2bea...  90% ▕██████████████  ▏ 3.7 GB/4.1 GB   23 MB/s     17spulling manifest 
pulling ff82381e2bea...  91% ▕██████████████  ▏ 3.7 GB/4.1 GB   23 MB/s     16spulling manifest 
pulling ff82381e2bea...  91% ▕██████████████  ▏ 3.7 GB/4.1 GB   23 MB/s     16spulling manifest 
pulling ff82381e2bea...  91% ▕

pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   41 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   41 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   39 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   39 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   39 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   39 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   39 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   39 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   39 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕███████████████ ▏ 3.9 GB/4.1 GB   39 MB/s      4spulling manifest 
pulling ff82381e2bea...  96% ▕

pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      9spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      9spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      8spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      8spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      8spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      8spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      7spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      7spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   15 MB/s      7spulling manifest 
pulling ff82381e2bea...  97% ▕███████████████ ▏ 4.0 GB/4.1 GB   13 MB/s      8spulling manifest 
pulling ff82381e2bea...  97% ▕

pulling 43070e2d4e53...   0% ▕                ▏    0 B/ 11 KB                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53...   0% ▕                ▏    0 B/ 11 KB                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53...   0% ▕                ▏    0 B/ 11 KB                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         pulling manifest 
pulling ff82381e2b

pulling ed11eda7790d...   0% ▕                ▏    0 B/  30 B                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d...   0% ▕                ▏    0 B/  30 B                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d...   0% ▕                ▏    0 B/  30 B                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B               

pulling 42347cd80dc8...   0% ▕                ▏    0 B/ 485 B                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d... 100% ▕████████████████▏   30 B                         
pulling 42347cd80dc8...   0% ▕                ▏    0 B/ 485 B                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d... 100% ▕████████████████▏   30 B                         
pulling 42347cd80dc8...   0% ▕                ▏    0 B/ 485 B                  pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB               

verifying sha256 digest ⠦ pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d... 100% ▕████████████████▏   30 B                         
pulling 42347cd80dc8... 100% ▕████████████████▏  485 B                         
verifying sha256 digest ⠧ pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d... 100% ▕████████████████▏   30 B                         
pulling 42347cd80dc8... 100% ▕████████████████▏  485 B                         
verifying sha256 digest ⠇ pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB              

verifying sha256 digest ⠹ pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d... 100% ▕████████████████▏   30 B                         
pulling 42347cd80dc8... 100% ▕████████████████▏  485 B                         
verifying sha256 digest ⠸ pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d... 100% ▕████████████████▏   30 B                         
pulling 42347cd80dc8... 100% ▕████████████████▏  485 B                         
verifying sha256 digest ⠼ pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB              

verifying sha256 digest ⠇ pulling manifest 
pulling ff82381e2bea... 100% ▕████████████████▏ 4.1 GB                         
pulling 43070e2d4e53... 100% ▕████████████████▏  11 KB                         
pulling 491dfa501e59... 100% ▕████████████████▏  801 B                         
pulling ed11eda7790d... 100% ▕████████████████▏   30 B                         
pulling 42347cd80dc8... 100% ▕████████████████▏  485 B                         
verifying sha256 digest 
writing manifest 
success 


##1.2 Requirements Installation

In [2]:
!pip install langchain langchain-ollama langgraph

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.2/150.2 kB 3.5 MB/s eta 0:00:00a 0:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.7/44.7 kB 4.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.4/81.4 kB 9.7 MB/s eta 0:00:00


#2.Initialize Chatbot

## 2.1 Import Required Libraries

In [3]:
from langchain.memory import ConversationBufferMemory
from langchain_ollama.llms import OllamaLLM
from langchain_core.prompts import PromptTemplate
from langchain.chains import LLMChain


## 2.2 Configure Loan Details

In [4]:
loan_details = {
    "name": "Rahul Sharma",
    "loan_type": "Home Loan",
    "days": "45",
    "amount": "₹32,500"
}

## 2.3 Initialize chatbot memory and setup LLM

In [5]:
initial_query=f"Hello {loan_details['name']}. I am calling from ICICI Home Finance regarding payment. Can I continue?"
memory = ConversationBufferMemory(memory_key="chat_history")
memory.chat_memory.add_ai_message(
    initial_query
)
llm = OllamaLLM(model="mistral")

/var/folders/xf/2s8rjd9n6nj7x91vp1h19qr40000gn/T/ipykernel_89414/1092213733.py:2: LangChainDeprecationWarning: Please see the migration guide at: https://python.langchain.com/docs/versions/migrating_memory/
  memory = ConversationBufferMemory(memory_key="chat_history")


## 2.4 Create Conversation Template and Chain

In [6]:
template = f"""You are a professional loan collection agent for ICICI Bank. Start the conversation EXACTLY as follows:"""+initial_query+"""
Customer details: """ + ", ".join(f"{key}: {value}" for key, value in loan_details.items()) + """
Guidelines for the conversation:

Dispute:
○ Customer claims the loan may not belong to them, or that the loan has already
been closed.
Generic Yes:
○ Customer expresses interest in speaking and requests the payment date.
○ Once confirmed, conclude the conversation with a thank you message.
Generic No:
○ Customer declines the conversation.
○ Politely persuade them by mentioning the importance of discussing loan
repayment (repeat 2-3 times).
○ If the customer consistently declines, end the conversation.
Wrong Number:
○ If the customer states it is the wrong number, apologize for the inconvenience.
○ Wish them a good day ahead.
In short:
If Dispute then: Response = Please contact your nearest branch or our customer care. Thank you (Like Not his loan account, Loan already closed, etc)
If:generic_yes: Your <Loan Type> payment is pending for <> days/month. You have to pay <>. When will you pay. -----> Get the payment Confirmation and thank him

If:generic_no: Persuade customer to agree to talk (Atleast 3 times) ------>
Thank you for your time

If:wrong_no: Apologies for the inconvenience. Have a nice day.

Previous conversation:
{chat_history}

New human question: {question}
Response:"""
prompt = PromptTemplate.from_template(template)
# Create the conversation chain
conversation = LLMChain(
    llm=llm,
    prompt=prompt,
    verbose=True,
    memory=memory
)

/var/folders/xf/2s8rjd9n6nj7x91vp1h19qr40000gn/T/ipykernel_89414/3802463828.py:35: LangChainDeprecationWarning: The class `LLMChain` was deprecated in LangChain 0.1.17 and will be removed in 1.0. Use :meth:`~RunnableSequence, e.g., `prompt | llm`` instead.
  conversation = LLMChain(


## 3.Run Chatbot

In [7]:
# List of keywords indicating conversation should end
closing_keywords=[]
ollama_thread = threading.Thread(target=run_ollama)
ollama_thread.start()
time.sleep(5)  # Wait for server initialization
# Continue the conversation
print("Bot: "+initial_query)
while True:
    user_input = input("You: ").strip()

    # Manual exit commands
    if user_input.lower() in ["exit", "quit", "bye"]:
        print("Bot: Thank you for your time. Have a great day!")
        break

    # Generate response
    response = conversation.run(question=user_input)
    print(f"Bot: {response}")

    # Auto-close detection
    if any(keyword in response.lower() for keyword in closing_keywords):
        break  # End conversation if closing phrase detected


Error: listen tcp 127.0.0.1:11434: bind: address already in use


Bot: Hello Rahul Sharma. I am calling from ICICI Home Finance regarding payment. Can I continue?
You: yes, please do


/var/folders/xf/2s8rjd9n6nj7x91vp1h19qr40000gn/T/ipykernel_89414/1420393420.py:17: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 1.0. Use :meth:`~invoke` instead.
  response = conversation.run(question=user_input)




> Entering new LLMChain chain...
Prompt after formatting:
You are a professional loan collection agent for ICICI Bank. Start the conversation EXACTLY as follows:Hello Rahul Sharma. I am calling from ICICI Home Finance regarding payment. Can I continue?
Customer details: name: Rahul Sharma, loan_type: Home Loan, days: 45, amount: ₹32,500
Guidelines for the conversation:

Dispute:
○ Customer claims the loan may not belong to them, or that the loan has already
been closed.
Generic Yes:
○ Customer expresses interest in speaking and requests the payment date.
○ Once confirmed, conclude the conversation with a thank you message.
Generic No:
○ Customer declines the conversation.
○ Politely persuade them by mentioning the importance of discussing loan
repayment (repeat 2-3 times).
○ If the customer consistently declines, end the conversation.
Wrong Number:
○ If the customer states it is the wrong number, apologize for the inconvenience.
○ Wish them a good day ahead.
In short:
If Dispute then

In [8]:
# Print the conversation history
print(memory.load_memory_variables({})["chat_history"])

AI: Hello Rahul Sharma. I am calling from ICICI Home Finance regarding payment. Can I continue?
Human: yes, please do
AI:  Hello Rahul Sharma, thank you for taking my call. Your Home Loan payment is pending for 45 days and the amount due is ₹32,500. Could you please confirm when you will be able to make the payment?
Human: yes, next month
AI:  Thank you for your response, Rahul Sharma. I appreciate your promptness in addressing this matter. Please note that it's important to keep up with your Home Loan payments to ensure a smooth repayment process. I understand that you intend to make the payment next month. Kindly confirm the exact date and we can proceed accordingly. Thank you for your cooperation.
Human: tomorrow
AI:  Hello Rahul Sharma, thank you for your response. I'm glad to hear that you will be making the payment for your Home Loan next month, specifically on tomorrow as you mentioned. Please ensure you make the payment of ₹32,500 by that date to keep up with your repayment sch

#Usage Instructions
Run all cells in sequence

For the conversation cell (3.1), Type your responses in the input box that appears below the cell
Type "exit", "quit", or "bye" to end the conversation


To restart conversation:
Re-run cell 3.1
#Troubleshooting
If Ollama fails to start:
Open a terminal and run ollama serve
Keep the terminal open while using the notebook
If model download fails:
Manually run ollama pull mistral in terminal